crates/sql_provider_datafusion/src/dbconnection.rs

```diff
@@ -10,14 +10,33 @@ pub mod postgresconn;
 pub type Error = Box<dyn std::error::Error + Send + Sync>;
 type Result<T, E = Error> = std::result::Result<T, E>;
 
-pub trait DbConnection<T, P>: Send {
+pub trait SyncDbConnection<T, P>: DbConnection<T, P> {
     fn new(conn: T) -> Self
     where
         Self: Sized;
     fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef>;
     fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
     fn execute(&self, sql: &str, params: &[P]) -> Result<u64>;
+}
 
+#[async_trait::async_trait]
+pub trait AsyncDbConnection<T, P>: DbConnection<T, P> + Sync {
+    fn new(conn: T) -> Self
+    where
+        Self: Sized;
+    async fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef>;
+    async fn query_arrow(&self, sql: &str, params: &[P]) -> Result<SendableRecordBatchStream>;
+    async fn execute(&self, sql: &str, params: &[P]) -> Result<u64>;
+}
+
+pub trait DbConnection<T, P>: Send {
     fn as_any(&self) -> &dyn Any;
     fn as_any_mut(&mut self) -> &mut dyn Any;
+
+    fn as_sync(&self) -> Option<&dyn SyncDbConnection<T, P>> {
```

Comment:
This feels very clunky. Why not just make the trait always async? Isn't it possible to run sync code in an async context?

Reply:
Tried the approach. The issue is the duckdb rs is synced and all struct not implemented sync properly.

Reply:
I couldn't find a way that makes the synchronised pool + duckdb inner structs thread safe. The duckdb connection is implemented in a synchronised way, e.g. InnerConnection. I don't think there is way to patch it.

Reply:
See my other comment